File: apps/api/app/schemas/detection.py
from datetime import datetime
from uuid import UUID
from typing import Optional
from pydantic import BaseModel, ConfigDict, Field


class DetectionType(str):
    LOG_ANALYSIS = "log_analysis"
    REQUEST_ANALYSIS = "request_analysis"
    PATTERN_MATCHING = "pattern_matching"
    ANOMALY_DETECTION = "anomaly_detection"
    SIGNATURE_BASED = "signature_based"
    BEHAVIORAL_ANALYSIS = "behavioral_analysis"
    ML_BASED = "ml_based"


class DetectionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    episode_id: UUID
    attack_id: Optional[UUID] = None
    detected: bool
    detection_type: str
    confidence: float
    details: dict
    matched_patterns: list[str]
    false_positive: bool
    timestamp: datetime
    created_at: datetime


class DetectionListParams(BaseModel):
    episode_id: UUID
    attack_id: Optional[UUID] = None
    detected: Optional[bool] = None
    detection_type: Optional[str] = None
    date_from: Optional[datetime] = None
    date_to: Optional[datetime] = None
    page: int = Field(default=1, ge=1)
    limit: int = Field(default=50, ge=1, le=100)


class DetectionStats(BaseModel):
    total_detections: int = 0
    true_positives: int = 0
    false_positives: int = 0
    false_negatives: int = 0
    detection_rate: float = 0.0
    by_type: dict = Field(default_factory=dict)

File: packages/shared/src/schemas/detection.ts
import { z } from "zod";
import { UUIDSchema, DateTimeSchema } from "./common";

export const DetectionTypeSchema = z.enum([
  "log_analysis",
  "request_analysis",
  "pattern_matching",
  "anomaly_detection",
  "signature_based",
  "behavioral_analysis",
  "ml_based"
]);
export type DetectionType = z.infer<typeof DetectionTypeSchema>;

export const DetectionSchema = z.object({
  id: UUIDSchema,
  episodeId: UUIDSchema,
  attackId: UUIDSchema.optional(),
  detected: z.boolean(),
  detectionType: DetectionTypeSchema,
  confidence: z.number().min(0).max(1),
  details: z.record(z.unknown()).default({}),
  matchedPatterns: z.array(z.string()).default([]),
  falsePositive: z.boolean().default(false),
  timestamp: DateTimeSchema,
  createdAt: DateTimeSchema
});
export type Detection = z.infer<typeof DetectionSchema>;

export const DetectionListParamsSchema = z.object({
  episodeId: UUIDSchema,
  attackId: UUIDSchema.optional(),
  detected: z.boolean().optional(),
  detectionType: DetectionTypeSchema.optional(),
  dateFrom: DateTimeSchema.optional(),
  dateTo: DateTimeSchema.optional(),
  page: z.number().int().positive().default(1),
  limit: z.number().int().positive().max(100).default(50)
});
export type DetectionListParams = z.infer<typeof DetectionListParamsSchema>;

export const DetectionStatsSchema = z.object({
  totalDetections: z.number().int().nonnegative(),
  truePositives: z.number().int().nonnegative(),
  falsePositives: z.number().int().nonnegative(),
  falseNegatives: z.number().int().nonnegative(),
  detectionRate: z.number().min(0).max(1),
  byType: z.record(z.object({
    total: z.number().int().nonnegative(),
    truePositives: z.number().int().nonnegative()
  }))
});
export type DetectionStats = z.infer<typeof DetectionStatsSchema>;